middleware/commandhandler/scheduler/middleware.go

```diff
@@ -82,6 +52,302 @@ func (c *command) ExecuteAt() time.Time {
 	return c.t
 }
 
+// NewMiddleware returns a new command handler middleware and a scheduler helper.
+func NewMiddleware(repo eh.ReadWriteRepo, codec eh.CommandCodec) (eh.CommandHandlerMiddleware, *Scheduler) {
+	s := &Scheduler{
+		repo:             repo,
+		cmdCh:            make(chan *scheduledCommand, 100),
+		cancelScheduling: map[uuid.UUID]chan struct{}{},
+		errCh:            make(chan error, 100),
+		codec:            codec,
+	}
+
+	return eh.CommandHandlerMiddleware(func(h eh.CommandHandler) eh.CommandHandler {
+		s.setHandler(h)
+
+		return eh.CommandHandlerFunc(func(ctx context.Context, cmd eh.Command) error {
+			// Delayed command execution if there is time set.
+			if c, ok := cmd.(Command); ok && !c.ExecuteAt().IsZero() {
+				// Use the wrapped command when created by the helper func.
+				innerCmd, ok := c.(*command)
+				if ok {
+					cmd = innerCmd.Command
+				}
+
+				// Ignore the persisted command ID in this case.
+				_, err := s.ScheduleCommand(ctx, cmd, c.ExecuteAt())
+
+				return err
+			}
+
+			// Immediate command execution.
+			return h.HandleCommand(ctx, cmd)
+		})
+	}), s
+}
+
+// PersistedCommand is a persisted command.
+type PersistedCommand struct {
+	ID         uuid.UUID       `json:"_"                 bson:"_id"`
+	IDStr      string          `json:"id"                bson:"_"`
+	RawCommand []byte          `json:"command"           bson:"command"`
+	ExecuteAt  time.Time       `json:"timestamp"         bson:"timestamp"`
+	Command    eh.Command      `json:"-"                 bson:"-"`
+	Context    context.Context `json:"-"                 bson:"-"`
+}
+
+// EntityID implements the EntityID method of the eventhorizon.Entity interface.
+func (c *PersistedCommand) EntityID() uuid.UUID {
+	return c.ID
+}
+
+// Scheduler is a scheduled of commands.
+type Scheduler struct {
+	h                  eh.CommandHandler
+	hMu                sync.Mutex
+	repo               eh.ReadWriteRepo
```

Comment:
Maybe a smaller interface can be used here?

If the decision to introduce a lock before handling the command, this repo-interface should include this.

Reply:
Good suggestion, will be handled in #364.